apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs

```diff
@@ -0,0 +1,311 @@
+    ///   session.ReleaseToPool();
+    /// If you use the SessionPool, you must make sure sessions are properly released back into the pool and are not simply GC'd.
+    /// Allowing a session to GC will incur penalties on the current process as the session will count towards the maximum allowed
+    /// and it will incur a penalty on other spanner processes because it will be an hour before the server frees the session if 
+    /// its not properly deleted.
+    /// </summary>
+    public static class SessionPool
+    {
+        private static readonly TimeSpan s_shutDownTimer = TimeSpan.FromSeconds(60);
+
+        // This member holds information we'll use when the session later gets released.
+        private static readonly ConcurrentDictionary<Session, SessionPoolKey>
+            s_sessionsInUse = new ConcurrentDictionary<Session, SessionPoolKey>();
+
+        private static int s_sessionsCreating;
+        private static readonly ConcurrentQueue<TaskCompletionSource<int>> s_waitQueue = new ConcurrentQueue<TaskCompletionSource<int>>();
+        private static readonly object s_waitSync = new object();
+
+        private static ConcurrentDictionary<SessionPoolKey, SessionPoolImpl> GlobalPool { get; } =
+            new ConcurrentDictionary<SessionPoolKey, SessionPoolImpl>();
+        private static readonly object s_priorityListSync = new object();
+        private static SortedList<SessionPoolImpl, SessionPoolImpl> PriorityList { get; }
+            = new SortedList<SessionPoolImpl, SessionPoolImpl>();
+
+        static SessionPool()
+        {
+#if NET45
+            AppDomain.CurrentDomain.DomainUnload += (sender, args) =>
+            {
+                Task.Run(ReleaseAll).Wait(s_shutDownTimer);
+            };
+#endif
+#if NETSTANDARD1_5
+            System.Runtime.Loader.AssemblyLoadContext.Default.Unloading += context =>
+            {
+                Task.Run(ReleaseAll).Wait(s_shutDownTimer);
+            };
+#endif
+        }
+
+        private static Task WhenCanceled(this CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSource<bool>();
+            cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).SetResult(true), tcs);
+            return tcs.Task;
+        }
+
+        private static async Task StartSessionCreating(CancellationToken cancellationToken)
+        {
+            lock (s_waitSync)
+            {
+                if (CurrentActiveSessions < MaximumActiveSessions)
+                {
+                    s_sessionsCreating++;
+                    return;
+                }
+            }
+            var canceledTask = cancellationToken.WhenCanceled();
+            //we need to block or throw because we are out of allocations.
+            while (true)
+            {
+                if (WaitOnResourcesExhausted)
+                {
+                    TaskCompletionSource<int> signal = new TaskCompletionSource<int>();
```

Comment:
Isn't there a problem that if a session was released between our check earlier and this part, we'll wait for a *second* session to be released unnecessarily, because we won't get told of the release? I suspect we want to make the check-and-register action atomic, effectively. I suspect I'd restructure this method significantly - let me know if you'd like me to sketch out some options instead of you doing so and then me critiquing them :)